extensions/amp-live-list/0.1/amp-live-list.js

```diff
@@ -0,0 +1,363 @@
+      item.classList.add(classes.ITEM);
+    });
+
+    this.validateLiveListItems_(this.itemsSlot_, true);
+
+    this.registerAction('update', this.update_.bind(this));
+  }
+
+  /** @override */
+  update(updatedElement) {
+    const container = this.getItemsSlot_(updatedElement);
+    this.validateLiveListItems_(container);
+    const mutateItems = this.getUpdates_(container);
+
+    // Insert/new items will be contiguous at the top even though they
+    // weren't in the actual request DOM structure.
+    mutateItems.insert.forEach(child => {
+      child.classList.add(classes.ITEM);
+      child.classList.add(classes.NEW_ITEM);
+      this.insertFragment_.insertBefore(child,
+          this.insertFragment_.firstElementChild);
+    });
+
+    if (mutateItems.insert.length > 0) {
+      const isHidden = this.updateSlot_
+          .classList.contains('-amp-hidden');
+      if (isHidden) {
+        this.updateSlot_.classList.remove('-amp-hidden');
+      }
+    }
+  }
+
+  /**
+   * Mutates the current elements dom and compensates for scroll
+   * change if necessary.
+   * @private
+   */
+  update_() {
+    if (this.insertFragment_.childElementCount == 0) {
+      return;
+    }
+
+    // TODO(erwinm): do in place update as well as sorting,
+    // correct insertion, tombstoning etc.
+    this.getVsync().mutate(() => {
+      // Remove the new class from the previously inserted items.
+      this.eachChildElement_(this.itemsSlot_, child => {
+        child.classList.remove(classes.NEW_ITEM);
+      });
+      this.itemsSlot_.insertBefore(this.insertFragment_,
+          this.itemsSlot_.firstElementChild);
+
+      // Hide the update button in case we previously displayed it.
+      this.updateSlot_.classList.add('-amp-hidden');
+      this.viewport_./*REVIEW*/scrollIntoView(this.element);
+    });
+  }
+
+  /** @override */
+  getInterval() {
+    return this.pollInterval_;
+  }
+
+  /**
+   * Seggregates new, updated and tombstoned elements.
+   *
+   * @param {!HTMLElement} element
+   * @return {!MutateItemsDef}
+   * @private
+   */
+  getUpdates_(updatedElement) {
+    // NOTE: We need to import the node for custom-element implementation
+    // to run. Reparenting directly or doing cloneNode alone won't work.
+    // We don't import the parent as importing the whole tree will
+    // trigger `createdCallback`s even though we won't actually be
+    // reparenting those nodes to this tree.
+    const insert = [];
+    const updates = [];
+    const tombstone = [];
+
+    for (let child = updatedElement.firstElementChild; child;
+        child = child.nextElementSibling) {
+      const id = child.getAttribute('id');
+
+      if (this.isChildNew_(child, id)) {
+        const orphan = this.win.document.importNode(child, true);
+        insert.unshift(orphan);
+        this.cacheChildId_(id, child.getAttribute('data-update-time'));
+        continue;
+      }
+
+      if (this.isChildUpdate_(child, id)) {
```

Comment:
It's not clear to me how the updates ever make it into main DOM. Please explain.